ast/compile_test.go

```diff
@@ -2125,6 +2142,72 @@ func TestCompilerCheckRuleConflictsDotsInRuleHeads(t *testing.T) {
 	}
 }
 
+func TestCompilerCheckRulePkgConflicts(t *testing.T) {
+	tests := []struct {
+		note    string
+		modules []*Module
+		err     []string
+	}{
+		{
+			note: "Package can be declared within dynamic extent of rule (#6387 regression test)",
+			modules: modules(
+				`package test
+					p[x] := y { x := "a"; y := "b" }`,
+				`package test.p
+					q := 1`),
+		},
+		{
+			note: "Package can be declared deep within dynamic extent of rule (#6387 regression test)",
+			modules: modules(
+				`package test
+					p[x] := y { x := "a"; y := "b" }`,
+				`package test.p.q.r.s
+					t := 1`),
+		},
+		{
+			note: "Package cannot be declared within extent of single-value rule (ground ref)",
+			modules: modules(
+				`package test
+					p := x { x := "a" }`,
+				`package test.p
+					q := 1`),
+			err: []string{
+				"rego_type_error: package test.p conflicts with rule p defined at mod0.rego:2",
+				"rego_type_error: rule data.test.p conflicts with [data.test.p.q]",
+			},
+		},
+		{
+			note: "Package cannot be declared within extent of multi-value rule",
+			modules: modules(
+				`package test
+					p[x] { x := "a" }`,
+				`package test.p
+					q := 1`),
+			err: []string{
+				"rego_type_error: package test.p conflicts with rule p defined at mod0.rego:2",
+				"rego_type_error: rule data.test.p conflicts with [data.test.p.q]",
```

Comment:
Unless there is an exception I'm not thinking of, we're always generating two errors for when a package is within the invalid extent of a rule. We could simply remove the pkg check completely, to drop the `package %s conflicts with rule %s` style error; but I'm hesitant to do so, as I think that one is more clear.

Reply:
Silencing the other error would be the better option, but that might prove difficult/messy.

Reply:
I would have expected only one error. I'm not familiar with the implementation details here so not sure if this is feasible.

Reply:
The same conflict is reported from "both ends". The problem is to know when to silence the second error, and when not to. Most of the time, it'll probably not get the second report.

I don't see it as a huge issue that we get two errors. You can solve the issue with the help of either, it's just unfortunate that the more clear error is the one that's trivial to remove.